ocaml/libs/tracing/tracing.ml

```diff
@@ -576,17 +576,70 @@ module Export = struct
 
       let set_trace_log_dir dir = trace_log_dir := dir
 
-      let export ~trace_id ~span_json ~path : (string, exn) result =
-        try
+      module FileStream = struct
+        let max_file_size = ref 1048576
+
+        let set_max_file_size size = max_file_size := size
+
+        let file_name = ref None
+
+        let file_stream = ref None
+
+        let lock = Mutex.create ()
+
+        let new_file_name path =
           let date = Ptime_clock.now () |> Ptime.to_rfc3339 ~frac_s:6 in
-          let file =
-            path
-            ^ String.concat "-" [trace_id; "xapi"; !host_id; date]
+          let ( // ) = Filename.concat in
+          let name =
+            (path // String.concat "-" [get_service_name (); !host_id; date])
             ^ ".json"
           in
-          Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file) 0o700 ;
-          Xapi_stdext_unix.Unixext.write_string_to_file file span_json ;
-          Ok ""
+          file_name := Some name ;
+          name
+
+        let open_stream file_name =
+          Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file_name) 0o700 ;
+          file_stream :=
+            Some (Unix.openfile file_name [O_WRONLY; O_CREAT; O_APPEND] 0o700)
+
+        let close_stream' () =
+          Option.iter (fun stream -> Unix.close stream) !file_stream ;
+          file_stream := None
+
+        let close_stream () =
+          Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
```

Comment:
Keep these in one function. The easy way to do this is:

```
Xapi_stdext_threads.Threadext.Mutex.execute lock @@ fun () ->
Option.iter (fun stream -> Unix.close stream) !file_stream ;
file_stream := None
```

Reply:
The reason this is being done is to avoid deadlock in `check` which is used to close a stream